primitives/src/block.rs

```diff
@@ -422,4 +428,144 @@ mod tests {
+    }
+
+    #[test]
+    fn block_hash_from_header_ref() {
+        let header = dummy_header();
+        let block_hash: BlockHash = BlockHash::from(&header);
+        assert_eq!(block_hash, header.block_hash());
+    }
+
+    #[test]
+    fn block_hash_from_block() {
+        let header = dummy_header();
+        let transactions = vec![];
+        let block = Block::new_unchecked(header, transactions);
+        let block_hash: BlockHash = BlockHash::from(block);
+        assert_eq!(block_hash, header.block_hash());
+    }
+
+    #[test]
+    fn block_hash_from_block_ref() {
+        let header = dummy_header();
+        let transactions = vec![];
+        let block = Block::new_unchecked(header, transactions);
+        let block_hash: BlockHash = BlockHash::from(&block);
+        assert_eq!(block_hash, header.block_hash());
+    }
+
+    #[test]
+    fn header_debug() {
+        let header = dummy_header();
+        let expected = format!(
+            "Header {{ block_hash: {:?}, version: {:?}, prev_blockhash: {:?}, merkle_root: {:?}, time: {:?}, bits: {:?}, nonce: {:?} }}",
+            header.block_hash(),
+            header.version,
+            header.prev_blockhash,
+            header.merkle_root,
+            header.time,
+            header.bits,
+            header.nonce
+        );
+        assert_eq!(format!("{:?}", header), expected);
+    }
+
+    #[cfg(feature = "arbitrary")]
+    #[cfg(feature = "alloc")]
+    #[test]
+    fn arbitrary_block() {
+        let data = vec![1u8; 100];
+        let mut unstructured = Unstructured::new(&data);
+
+        let block = Block::arbitrary(&mut unstructured).unwrap();
+
+        println!("{:?}", block);
+        let header = block.header;
+        assert!(header.version.to_consensus() > 0);
+        assert!(header.prev_blockhash.as_byte_array().len() == 32);
+        assert!(header.merkle_root.as_byte_array().len() == 32);
+        assert!(header.version.to_consensus() > 0);
+    }
+
+    #[cfg(feature = "arbitrary")]
+    #[test]
+    fn arbitrary_version_branches() {
+        let data = vec![0u8; 100];
+        let mut unstructured = Unstructured::new(&data);
+
+        // Test branch where choice is 0
+        let version = Version::arbitrary(&mut unstructured).unwrap();
+        assert_eq!(version, Version::ONE);
+
+        let data = vec![1u8; 100];
+        let mut unstructured = Unstructured::new(&data);
+        let version = Version::arbitrary(&mut unstructured).unwrap();
+        assert_eq!(version, Version::TWO);
+
+        let data = vec![2u8; 100];
+        let mut unstructured = Unstructured::new(&data);
+        let version = Version::arbitrary(&mut unstructured).unwrap();
+        assert_eq!(version, Version::NO_SOFT_FORK_SIGNALLING);
+
+        let data = vec![3u8; 100];
+        let mut unstructured = Unstructured::new(&data);
+        let version = Version::arbitrary(&mut unstructured).unwrap();
+        assert!(version.to_consensus() >= 0);
+    }
```

Comment:
This doesn't seem correct. The entire point of `arbitrary` is to run many iterations with various randomly-generated values. Usually people use external crates for that.

Reply:
Yeah true. The tests were just to check that the arbitrary functions themselves work.  I could write another test that actually uses them properly and checks a series of random values.  This does seem like fuzzing though, and maybe instead primitives should be added to the crates that the fuzz tests are done on?

Reply:
> I could write another test that actually uses them properly and checks a series of random values. This does seem like fuzzing though, and maybe instead primitives should be added to the crates that the fuzz tests are done on?

Yes, I agree.  A good way to make sure the Arbitrary types are usable is to use them in fuzz testing and proptesting.  Although, I do see some value in having these tests that show the Arbitrary types behave as designed.

Reply:
> The tests were just to check that the arbitrary functions themselves work.

And this, right here is an example why I think coverage shouldn't be used as a religion. The point of `arbitrary` is to generate *arbitrary* values. The exact algorithm is not specified and there's no reason to try to define it. If there's anything regarding `arbitrary` that perhaps should be tested it's that the function itself doesn't panic on any seed.

Reply:
> that perhaps should be tested it's that the function itself doesn't panic on any seed.

I think if they panic that's discoverable when using them in implementation testing and then can be fixed.  Not to mention, how would you test not panicking on any seed without giving it every seed :thinking: .

Reply:
> I think if they panic that's discoverable when using them in implementation testing and then can be fixed.

Well but if it happens in downstream code it'll be an annoying bug since it'll prevent downstream from testing until it's fixed. But yes, if *we* have any test that uses it then it's already tested.

> how would you test not panicking on any seed without giving it every seed

By feeding it a bunch of inputs we can at least make the assumption that it's unlikely. But also `kani` can almost to it.

Reply:
I agree that the arbitrary tests I added don't fully test the arbitrary implementations.  But at the same time I don't think they make the tests worse.  And while test coverage isn't everything these tests do bring it to 100% which makes it clear if new code is added that is not covered by tests.

Adding further tests or fuzzing is probably a good idea, but does that also mean these two tests should be removed?

Reply:
> Adding further tests or fuzzing is probably a good idea, but does that also mean these two tests should be removed?

The question I have is do we want to add this and maintain in for every arbitrary type.